singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,377 @@
+  // initializing with pressure coeffs to get energy coeffs into VIP
+  _VIP[2] = 1.0;                                                // prefactor d0
+  _VIP[3] = 0.0;                                                // prefactor d1
+  for (int ind = 4; ind < VinetInternalParametersSize; ind++) { //_d2tod40[0]=d2
+    _VIP[ind] = _d2tod40[ind - 4];                              // dn is in VIP[n+2]
+  }
+  for (int ind = VinetInternalParametersSize - 2; ind >= 2;
+       ind--) { // _VIP[42]=d40=f40 given,first calculated is _VIP[41]=f39
+    _VIP[ind] = _VIP[ind] - (ind) / _VIP[1] * _VIP[ind + 1]; // prefactors f40 to f0
+  }                                                          // _VIP[n+2]=fn, ind=n+2
+}
+
+PORTABLE_INLINE_FUNCTION void Vinet::Vinet_F_DT_func(const Real rho, const Real T,
+                                                     Real *output) const {
+  int pref0vp = -2, pref0vip = 2, maxind = VinetInternalParametersSize - 3;
+  Real sumP = 0.0, sumB = 0.0, sumE = 0.0;
+
+  Real x = pow(_rho0 / rho, 1.0 / 3.0); /*rho-dependent*/
+  Real x2inv = 1.0 / pow(x, 2.0);
+  Real onemx = 1.0 - x;
+  Real etatimes1mx = _VIP[1] * onemx;
+  Real expetatimes1mx = exp(etatimes1mx);
+
+  for (int ind = maxind; ind >= 2; ind--) {              //_d2tod40[0]=d2
+    sumP = _d2tod40[pref0vp + ind] + onemx * sumP;       //_d2tod40[38]=d40
+    sumB = _d2tod40[pref0vp + ind] * ind + onemx * sumB; //_d2tod40[-2+40]=d40
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;          //_VIP[42]=f40
+  }                                                      //_VIP[2]=f0
+  for (int ind = 1; ind >= 0; ind--) {
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;
+  }
+  sumP = 1.0 + sumP * pow(onemx, 2.0);
+  sumB = sumB * onemx;
+  sumE = _VIP[1] * onemx * sumE;
+
+  /* T0 isotherm */
+  Real energy = 9.0 * _B0 / pow(_VIP[1], 2.0) / _rho0 *
+                    (_VIP[pref0vip] - expetatimes1mx * (_VIP[pref0vip] - sumE)) -
+                _A0 * _B0 * _T0 * (1.0 / _rho0 - 1.0 / rho);
+  Real pressure = 3.0 * _B0 * x2inv * onemx * expetatimes1mx * sumP;
+  Real temp = (1.0 + onemx * (_VIP[1] * x + 1.0)) * sumP + x * onemx * sumB;
+  temp = _B0 / _rho0 * x * expetatimes1mx * temp;
+
+  /* Go to required temperature */
+  energy = energy + _Cv0 * (T - _T0) + _E0;
+  pressure = pressure + _A0 * _B0 * (T - _T0);
+  Real dpdrho = temp;
+  Real dpdt = _A0 * _B0;
+  Real dedt = _Cv0;
+  Real dedrho = (pressure - T * _A0 * _B0) / pow(rho, 2.0);
+  Real entropy;
+  if (T < 0.0) {
+    printf("warning!! negative temperature");
+    entropy = 0.0;
+  } else {
+    entropy = _A0 * _B0 * (1.0 / rho - 1 / _rho0) + _Cv0 * log(T / _T0) + _S0;
+  }
+  Real soundspeed = std::max(0.0, dpdrho + T * pow(dpdt / rho, 2.0) / _Cv0);
+  soundspeed = sqrt(soundspeed);
+
+  output[0] = energy;
+  output[1] = pressure;
+  output[2] = dpdrho;
+  output[3] = dpdt;
+  output[4] = dedt;
+  output[5] = dedrho;
+  output[6] = entropy;
+  output[7] = soundspeed;
```

Comment:
It might be useful to give these indices names by constants such as `constexpr int IE = 0` (i.e., energy index = 0) as fields in the class, so you have access to them everywhere and don't have to remember that 0 is energy, 1 is pressure, etc.

Reply:
Is there a reason you don't want to add the named variables mapping to these indices @aematts ? It would allow you to reference them in your functions below which would simplify things.